### PR TITLE
GSSAPI support via PECL krb5 module

### PR DESCRIPTION
**Pull request by urusha**
Not a php specialist, sure code could be written smarter. While this works for me, may be this should be applied too: https://github.com/roundcube/roundcubemail/issues/6308

### CODE REVIEW

File: Net/SMTP.php [hunk 1]
```diff
@@ -198,6 +202,7 @@ public function __construct($host = null, $port = null, $localhost = null,
         /* These standard authentication methods are always available. */
         $this->setAuthMethod('LOGIN', array($this, 'authLogin'), false);
         $this->setAuthMethod('PLAIN', array($this, 'authPlain'), false);
+        $this->setAuthMethod('GSSAPI', array($this, 'authGSSAPI'), false);
```
Comment (alecpl) [inline Net/SMTP.php]: I would register the method first, and do it only if principal and cname is set. Maybe checking for krb5 extension should be done here (as we do with checking SASL class).

File: Net/SMTP.php [hunk 2]
```diff
@@ -871,6 +876,95 @@ protected function authPlain($uid, $pwd, $authz = '')
         return true;
     }
 
+        /**
```
Comment (alecpl) [inline Net/SMTP.php]: Looks like an indentation issue.

File: Net/SMTP.php [hunk 3]
```diff
@@ -871,6 +876,95 @@ protected function authPlain($uid, $pwd, $authz = '')
...
+     * @note PHP krb5 extension is required and the service principal and
+     *       credentials cache must have been set.
+     * @param string $uid   The userid to authenticate as.
+     * @param string $pwd   The password to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     */
+
```
Comment (alecpl) [inline Net/SMTP.php]: Redundant empty line.

File: Net/SMTP.php [hunk 4]
```diff
@@ -871,6 +876,95 @@ protected function authPlain($uid, $pwd, $authz = '')
...
+            return $error;
+        }
+
+        /* 235: Authentication successful */
+        if (PEAR::isError($error = $this->parseResponse(235))) {
+            return $error;
+        }
+
+        return true;
+
```
Comment (alecpl) [inline Net/SMTP.php]: Redundant empty line.

File: Net/SMTP.php [hunk 5]
```diff
@@ -871,6 +876,95 @@ protected function authPlain($uid, $pwd, $authz = '')
         return true;
     }
 
+        /**
+     * Authenticates the user using the GSSAPI method.
+     *
+     * @note PHP krb5 extension is required and the service principal and
```
Comment (alecpl) [inline Net/SMTP.php]: No such thing as @note in phpdoc, I think. Leave the text but just remove the `@note ` prefix.

**Comment by urusha** [comment]
Should I squash commits?

**Comment by alecpl** [comment]
This is similar code to what we already have in Net_Sieve. Looks good to me.
@jparise @ashnazg could we have it merged and released soon?

**Comment by urusha** [comment]
Seems to be done.

File: docs/guide.txt [hunk 6]
```diff
@@ -67,12 +67,29 @@ methods, in order of preference:
...
+.. _RFC-4120: https://tools.ietf.org/html/rfc4120
+.. _krb5: https://pecl.php.net/package/krb5
+
 DIGEST-MD5
 ----------
 
 The DIGEST-MD5 authentication method uses `RSA Data Security Inc.`_'s MD5
-Message Digest algorithm.  It is considered the most secure method of SMTP
-authentication.
+Message Digest algorithm.  It is considered more secure method of SMTP
```
Comment (jparise) [inline docs/guide.txt]: "a" more secure method?

**Comment by jparise** [comment]
Looks good. Thanks @urusha!

**Pull request merged**